### PR TITLE
Smaller fixes on new catchall and userchange

### PR DESCRIPTION
**Pull request by Udera**
Some smaller fixes:

* New catchall, we used `$_POST[user_id]` even though this does not exist
* userchange: when only the `realname` is changed in the upper form, the script runs the other updates and creates errors on non existing `$_POST[value]`-variables.
* userchange: option to `store forwarded mail locally` makes only sense when forwarding is active.

### CODE REVIEW

**Comment by rimas-kudelis** [comment]
I'm not entirely sure about these:
> * userchange: when only the `realname` is changed in the upper form, the script runs the other updates and creates errors on non existing `$_POST[value]`-variables.

Even if you change just the real name, you're still submitting the whole form. If there are unprevented errors due to missing variables, these errors should be fixed instead. 

> * userchange: option to `store forwarded mail locally` makes only sense when forwarding is active.

Then the checkbox in question should be disabled and possibly hidden when forwarding is inactive. But I don't think we should be ignoring changes in this value just because some other value means this one will not be used. Don't you think so?

File: vexim/userchangesubmit.php [hunk 1]
```diff
@@ -28,7 +28,7 @@
     }
   }
 
-  if ($_POST['realname'] !== "") {
+  if (isset($_POST['realname'])) {
```
Comment (rimas-kudelis) [inline vexim/userchangesubmit.php]: IMO, checking for non-empty realname was a good thing to do. Perhaps you should use both `isset()` and `!= ''`?

**Comment by Udera** [comment]
> Even if you change just the real name, you're still submitting the whole form. If there are unprevented errors due to missing variables, these errors should be fixed instead.

there are actually 3 different forms.


> Then the checkbox in question should be disabled and possibly hidden when forwarding is inactive. But I don't think we should be ignoring changes in this value just because some other value means this one will not be used. Don't you think so?

I check the vexim-routers, and the `unseen`-option only works if vacation is on. Then we can keep the existing settings. To show this option only when vacation is active, that's probably something for javascript.

**Comment by rimas-kudelis** [comment]
> there are actually 3 different forms.

OK. But in fact, the right way to update the user would be via a single `UPDATE` query, not three different queries run under different circumstances, which require tight coordination between how frontend forms and backend processing are made. But that's probably one of the things to do for Vexim3?

> I check the vexim-routers, and the unseen-option only works if vacation is on. Then we can keep the existing settings. To show this option only when vacation is active, that's probably something for javascript.

Of course we *can* keep the existing settings, but the question is whether we *should*. I think we should respect the changes, if they are being made.
And yes, disabling of the dependent field will have to be done via javascript.

**Comment by Udera** [comment]
>> there are actually 3 different forms.
> 
> OK. But in fact, the right way to update the user would be via a single UPDATE query, not three different queries run under different circumstances, which require tight coordination between how frontend forms and backend processing are made. But that's probably one of the things to do for Vexim3?

We have this problem on nearly every page in current vexim. In some scripts, it would change the logic, so I don't think it's worth to consider such changes in this version. In the new version there is hopefully a more responsive design.



>> I check the vexim-routers, and the unseen-option only works if vacation is on. Then we can keep the existing settings. To show this option only when vacation is active, that's probably something for javascript.
> 
> Of course we can keep the existing settings, but the question is whether we should. I think we should respect the changes, if they are being made.
> And yes, disabling of the dependent field will have to be done via javascript.

The existing setting makes sense: Someone forwards his mail during vacation and keeps a local copy. The next time he activates the forwarder he wants to use the same settings.

The javascript to hide the features unless forwarding is activated could be useful at a number of other pages (e.g. spamassassin points).

**Comment by rimas-kudelis** [comment]
> The existing setting makes sense: Someone forwards his mail during vacation and keeps a local copy. The next time he activates the forwarder he wants to use the same settings.

I still don't understand why we should ignore changes the users make just because these changes cannot have any immediate effect. The only explanation for that that I could come up with is "it makes it more obvious whether forwarding is on or off", but I'm not sure if that is a rationale good enough.

Anyway, looking at `userchangesubmit.php` again, now I see we're doing the same with forward e-mail address, which we also clear when forwarding is off. In that case I suppose we can keep this behavior for now. Would you mind merging your `if` block with the `if` block above?

**Comment by Udera** [comment]
I harmonized the last submit-button and fixed the javascript to focus the first input field as well.

**Comment by rimas-kudelis** [comment]
Yeah, but... Would you mind merging the two `if` blocks in `userchangesubmit.php` so that we don't check the value of `$_POST['on_forward']` twice?

**Comment by Udera** [comment]
Now, I got the `if`-block you were referring to (hopefully) ;-)

**Comment by rimas-kudelis** [comment]
Yep, that's the block I was talking about. But you forgot to paste the `if (isset($_POST['unseen'])) {$_POST['unseen'] = 1;}` piece. :)

**Comment by Udera** [comment]
I was too happy, to find the right place that I forgot about the initial purpose :D

**Comment by rimas-kudelis** [comment]
Am I **that** cryptic with my comments? o_O

**Comment by Udera** [comment]
It's too early, there are too many `if`s, ...

**Comment by rimas-kudelis** [comment]
Would you please put `$_POST['unseen'] = 1;` on its own line, indent it prorperly and add curly braces around it?

**Comment by Udera** [comment]
voilà, with curly brackets :D

**Comment by rimas-kudelis** [comment]
Come to think of it, we should set `unseen` to `0` if it is not set, but forwarding is enabled... I forgot about that before. :/

**Comment by rimas-kudelis** [comment]
I suppose this little `if ` could even be replaced with this:
`$_POST['unseen'] = (int)isset($_POST['unseen'])`

And above that block, there's this piece of code currently:
`if (isset($_POST['on_vacation'])) {$_POST['on_vacation'] = 1;} else {$_POST['on_vacation'] = 0;}`
which could then be replaced with this:
`$_POST['on_vacation'] = (int)isset($_POST['on_vacation'])`

Although in future versions we should probably allow flags to be set to `0` meaning "off".

**Comment by rimas-kudelis** [comment]
Alternative syntax (even better):
`$_POST['unseen'] = isset($_POST['unseen']) ? 1 : 0;`
and
`$_POST['on_vacation'] = isset($_POST['on_vacation']) ? 1 : 0;`

**Comment by Udera** [comment]
If you have more ideas, I'm back online tonight ;-)

**Comment by rimas-kudelis** [comment]
Definitely not in this PR. :)

Thank you once more, I really appreciate your patience!

**Pull request merged**